File: src/stages/stage0.ts
import fs from "node:fs/promises";
import path from "node:path";

import type { AuditConfig } from "../config.js";
import { getLogger } from "../logger.js";

const logger = getLogger("stage0");

export async function runSetup(config: AuditConfig): Promise<void> {
  const directories = [
    config.outputDir,
    path.join(config.outputDir, ".markers"),
    path.join(config.outputDir, "stage-1-details"),
    path.join(config.outputDir, "stage-3-details"),
    path.join(config.outputDir, "stage-4-details"),
    path.join(config.outputDir, "stage-5-details"),
    path.join(config.outputDir, "stage-5-details", "_pending"),
  ];

  for (const directory of directories) {
    await fs.mkdir(directory, { recursive: true });
    logger.debug("Directory ready: %s", directory);
  }

  logger.info("Stage 0 complete. Output dir: %s", config.outputDir);
}

File: code_auditor/stages/stage0.py
from __future__ import annotations

import os
import subprocess

from ..config import AuditConfig
from ..logger import get_logger

logger = get_logger("stage0")


def _is_git_repo(path: str) -> bool:
    return os.path.isdir(os.path.join(path, ".git"))


def _git_pull(target: str) -> None:
    """Stash uncommitted changes if any, pull latest, then restore the stash."""
    logger.info("Target is a git repo. Pulling latest changes...")

    # Check for uncommitted changes (staged, unstaged, or untracked)
    status = subprocess.run(
        ["git", "status", "--porcelain"],
        cwd=target, capture_output=True, text=True, check=True,
    )
    has_changes = bool(status.stdout.strip())

    if has_changes:
        logger.info("Stashing uncommitted changes before pull.")
        subprocess.run(
            ["git", "stash", "--include-untracked"],
            cwd=target, capture_output=True, text=True, check=True,
        )

    try:
        result = subprocess.run(
            ["git", "pull"],
            cwd=target, capture_output=True, text=True, check=True,
        )
        logger.info("git pull: %s", result.stdout.strip() or "up to date")
    finally:
        if has_changes:
            logger.info("Restoring stashed changes.")
            try:
                subprocess.run(
                    ["git", "stash", "pop"],
                    cwd=target, capture_output=True, text=True, check=True,
                )
            except subprocess.CalledProcessError as e:
                logger.warning(
                    "Failed to restore stashed changes (merge conflict?): %s\n"
                    "stderr: %s\n"
                    "Your changes remain in the stash. Run 'git stash pop' manually to recover them.",
                    e, e.stderr.strip() if e.stderr else "",
                )


async def run_setup(config: AuditConfig) -> None:
    if _is_git_repo(config.target):
        _git_pull(config.target)

    directories = [
        config.output_dir,
        os.path.join(config.output_dir, ".markers"),
        os.path.join(config.output_dir, "stage1-security-context"),
        os.path.join(config.output_dir, "stage2-analysis-units"),
        os.path.join(config.output_dir, "stage3-findings"),
        os.path.join(config.output_dir, "stage4-vulnerabilities"),
        os.path.join(config.output_dir, "stage4-vulnerabilities", "_pending"),
        os.path.join(config.output_dir, "stage5-pocs"),
        os.path.join(config.output_dir, "stage6-disclosures"),
    ]

    for directory in directories:
        os.makedirs(directory, exist_ok=True)
        logger.debug("Directory ready: %s", directory)

    logger.info("Stage 0 complete. Output dir: %s", config.output_dir)
